pkg/storage/plugins/mongodb/mongodb.go

```diff
@@ -0,0 +1,255 @@
+package mongodb
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"get.porter.sh/porter/pkg/context"
+	"get.porter.sh/porter/pkg/storage/plugins"
+	"github.com/globalsign/mgo"
+	"github.com/globalsign/mgo/bson"
+	"github.com/pkg/errors"
+)
+
+var (
+	_               plugins.StoragePlugin = &Store{}
+	ErrNotConnected                       = errors.New("cannot execute command against the mongodb plugin because the session is closed (or was never connected)")
+)
+
+// Store implements the Porter plugin.StoragePlugin interface for mongodb.
+type Store struct {
+	*context.Context
+
+	url      string
+	database string
+	session  *mgo.Session
+}
+
+// NewStore creates a new storage engine that uses MongoDB.
+func NewStore(c *context.Context, cfg PluginConfig) *Store {
+	return &Store{
+		Context: c,
+		url:     cfg.URL,
+	}
+}
+
+func (s *Store) Connect() error {
+	if s.session != nil {
+		return nil
+	}
+
+	host, port, db, err := parseConnectionString(s.url)
+	if err != nil {
+		return err
+	}
+	if db == "" {
+		db = "porter"
+	}
+	s.database = db
+
+	if s.Debug {
+		fmt.Fprintf(s.Err, "Connecting to mongo database %s at %s on %s\n", s.database, host, port)
+	}
+	session, err := mgo.Dial(s.url)
+	if err != nil {
+		return err
+	}
+	s.session = session
+	return nil
+}
+
+func (s *Store) Close() error {
+	if s.session != nil {
+		s.session.Close()
+		s.session = nil
+	}
+	return nil
+}
+
+// Ping the connected session to check if everything is okay.
+func (s *Store) Ping() error {
+	session, err := s.copySession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	return session.Ping()
+}
+
+func (s *Store) Aggregate(opts plugins.AggregateOptions) ([]bson.Raw, error) {
+	session, err := s.copySession()
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	// TODO(carolynvs): wrap each call with session.refresh  on error and a single retry
+	c := s.getCollection(session, opts.Collection)
+	var results []bson.Raw
+	err = c.Pipe(opts.Pipeline).All(&results)
+	return results, err
+}
+
+// EnsureIndexes makes sure that the specified indexes exist and are
+// defined appropriately.
+func (s *Store) EnsureIndex(opts plugins.EnsureIndexOptions) error {
+	session, err := s.copySession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	c := s.getCollection(session, opts.Collection)
+	return c.EnsureIndex(opts.Index)
+}
+
+func (s *Store) Count(opts plugins.CountOptions) (int, error) {
+	session, err := s.copySession()
+	if err != nil {
+		return 0, err
+	}
+	defer session.Close()
+
+	c := s.getCollection(session, opts.Collection)
+	return c.Find(opts.Filter).Count()
+}
+
+func (s *Store) Find(opts plugins.FindOptions) ([]bson.Raw, error) {
+	session, err := s.copySession()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
```

Comment:
nit: just noticed this could probably be simplified to `defer session.Close()` like the others?